File: src/main/java/io/playground/orderservice/application/saga/port/persistence/ProcessSagaPersistencePort.java
package io.playground.orderservice.application.saga.port.persistence;

import io.playground.orderservice.domain.saga.ProcessSaga;

import java.time.Instant;
import java.util.List;
import java.util.Optional;

public interface ProcessSagaPersistencePort {
    Optional<ProcessSaga> findByOrderExternalId(String orderExternalId);

    ProcessSaga saveAndFlush(ProcessSaga processSaga);

    List<ProcessSaga> findExpiredSagas(int limitSize,
                                       int retryCount);

    boolean updateStatus(Long id,
                         ProcessSaga.ProcessSagaStatus status);

    boolean updateStatus(String orderExternalId,
                         ProcessSaga.ProcessSagaStatus status,
                         ProcessSaga.ProcessSagaStatus beforeStatus);

    boolean updatePaymentKey(String orderExternalId,
                             String paymentKey);

    boolean updateRetryCountAndLockedUntil(Long id,
                                           Instant lockedUntil);
}

File: src/main/kotlin/io/playground/orderservice/application/saga/port/persistence/ProcessSagaPersistencePort.kt
package io.playground.orderservice.application.saga.port.persistence

import io.playground.orderservice.domain.saga.ProcessSaga
import java.time.Instant
import java.util.Optional

interface ProcessSagaPersistencePort {
    fun findByOrderExternalId(orderExternalId: String): Optional<ProcessSaga>

    fun saveAndFlush(processSaga: ProcessSaga): ProcessSaga

    fun findExpiredSagas(limitSize: Int, retryCount: Int): List<ProcessSaga>

    fun updateStatus(id: Long, status: ProcessSaga.ProcessSagaStatus): Boolean

    fun updateStatus(
        orderExternalId: String,
        status: ProcessSaga.ProcessSagaStatus,
        beforeStatus: ProcessSaga.ProcessSagaStatus,
    ): Boolean

    fun updatePaymentKey(orderExternalId: String, paymentKey: String): Boolean

    fun updateRetryCountAndLockedUntil(id: Long, lockedUntil: Instant?): Boolean
}
